tap_file/streams.py

```diff
@@ -3,45 +3,151 @@
 from __future__ import annotations
 
 import csv
+import json
+import re
 from functools import cached_property
 from typing import Any, Generator
 
 from tap_file.client import FileStream
 
 
-class CSVStream(FileStream):
-    """Stream for reading CSVs."""
-
-    name = "CSV"
+class DelimitedStream(FileStream):
+    """Stream for reading CSVs and TSVs."""
 
     def get_rows(self) -> Generator[dict[str | Any, str | Any], None, None]:
-        """Retrive all rows from all CSVs.
+        """Retrive all rows from all *SVs.
 
         Yields:
-            A dictionary containing information about a row in a CSV.
+            A dictionary containing information about a row in a *SV.
         """
-        for file in self.get_files():
-            with self.filesystem.open(path=file, mode="rt") as f:
-                reader = csv.DictReader(f)
-                for row in reader:
-                    yield row
+        for reader in self._get_readers():
+            for row in reader:
+                yield row
 
     @cached_property
     def schema(self) -> dict[str, dict]:
-        """Create a schema for a CSV file.
+        """Create a schema for a *SV file.
 
-        Each column in the CSV will have its own entry in the schema. All entries will
+        Each column in the *SV will have its own entry in the schema. All entries will
         be of the form: `'FIELD_NAME': {'type': ['null', 'string']}`
 
         Returns:
-            A schema representing a CSV.
+            A schema representing a *SV.
         """
         properties = {}
 
+        for reader in self._get_readers():
+            for field in reader.fieldnames:
+                properties.update({field: {"type": ["null", "string"]}})
+
+        return {"properties": properties}
+
+    def _get_readers(self) -> Generator[csv.DictReader[str], None, None]:
+        quote_character: str = self.config["quote_character"]
+
+        for file in self.get_files():
+            if self.config["delimiter"] == "detect":
+                if re.match(".*\\.csv.*", file):
+                    delimiter = ","
+                elif re.match(".*\\.tsv.*", file):
+                    delimiter = "\t"
+                else:
+                    msg = (
+                        "Configuration option 'delimiter' is set to 'detect' but a "
+                        "non-csv non-tsv file is present. Please manually specify "
+                        "'delimiter'."
+                    )
+                    raise RuntimeError(msg)
+            else:
+                delimiter = self.config["delimiter"]
+
+            with self.filesystem.open(
```

Comment:
How does fsspec handle encodings? Does it default to utf8. We can make a seperate issue for this but I'm just remembering https://github.com/MeltanoLabs/tap-csv/blob/main/tap_csv/client.py#L117 and https://github.com/MeltanoLabs/tap-csv/pull/88

Reply:
Yes, it defaults to UTF8. https://filesystem-spec.readthedocs.io/en/latest/api.html?highlight=encoding:#fsspec.gui.FileSelector.open_file